cloudinit/url_helper.py

```diff
@@ -101,7 +101,7 @@ def read_file_or_url(url, timeout=5, retries=10,
             raise UrlError(cause=e, code=code, headers=None, url=url)
         return FileResponse(file_path, contents=contents)
     else:
-        return readurl(url, timeout=timeout, retries=retries,
+        return readurl(url, timeout=timeout, retries=retries, headers=headers,
```

Comment:
Would it make sense to modify the signature of `read_file_or_url` to take `*args` and `**kwargs` which we can then just pass through here?

This would future-proof us against regression (and the annoying work of keeping function signatures in sync).

Reply:
I like it. I have adapted this changeset. Let me know what you think

Reply:
I also like it, but that means you need to forbid the use of positional arguments other than `url`, and the language feature that enforces that isn't in the lowest versions of Python cloud-init supports. Some other enforcement mechanism would be required.

Reply:
This is a great point Jason, thanks for bringing it up! If we drop `*args` from the function signature then I think we actually can make it keyword-only in Python 2 (but only because we're using `**kwargs`):

```py
>>> def foo(url, **kwargs):
...   pass
... 
>>> foo("url")
>>> foo("url", 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes exactly 1 argument (2 given)
```

I'll comment on the appropriate line in the diff too.

Reply:
Let's not introduce more unexpected change. Please note that the previous version of read_file_or_url changed some of the default parameters for readurl.
```
def read_file_or_url(url, timeout=5, retries=10,
                     headers=None, data=None, sec_between=1, ssl_details=None,
                     headers_cb=None, exception_cb=None):
```
And here's the previous readurl signature:
```
def readurl(url, data=None, timeout=None, retries=0, sec_between=1,
            headers=None, headers_cb=None, ssl_details=None,
            check_status=True, allow_redirects=True, exception_cb=None,
            session=None, infinite=False, log_req_resp=True,
            request_method=None):
```
Switching read_file_or_url to a pure passthrough will lose these changed "defaults" which can easily break existing code that relies on the current behavior.

Reply:
That's a good catch on the differences in defaults, thanks! I think I would prefer to have consistent defaults for the two functions, rather than have the behaviour of a "read" change depending on which one is called.

> existing code that relies on the current behavior.

All the existing code is in this repository, so we can adjust it to be explicit about its different retry/timeout behaviour as part of this PR. Grepping through the (non-test) code, `read_file_or_url` is called 12 times, but 9 of those already pass values for `timeout` and `retries`. The three others are:

* `cloudinit/user_data.py:227`
* `cloudinit/sources/helpers/azure.py:186`
* `cloudinit/sources/helpers/azure.py:193`

In my opinion, being explicit about behaviour in three extra call sites is worth losing the inconsistency between these two `read` functions. What do you think?

Reply:
I agree with you that fixing the three "defaulting" call sites and re-specifying read_file_or_url to be explicitly a passthrough is the smart thing to do.

I'd really like to be able to change the readurl signature to enforce "url is required to be positional, all others are required to be keyword, there are no positional-or-keyword parameters", like so:
```
def readurl(url, /, *, data=None, timeout=None, retries=0, sec_between=1,
            headers=None, headers_cb=None, ssl_details=None,
            check_status=True, allow_redirects=True, exception_cb=None,
            session=None, infinite=False, log_req_resp=True,
            request_method=None):
```
Unfortunately, that syntax requires PEP 570, which is a Python 3.8 feature. Dropping the *args parameter from read_file_or_url blocks attempts to use positional parameters other than url, but still allows someone to pass url through a kwarg dict.

I'm sure this is just my "I'm actually a C++ programmer" knee jerking here; I keep trying to enforce as much compile-time checking as possible, and that's not really the Python way.

Reply:
> Unfortunately, that syntax requires PEP 570, which is a Python 3.8 feature.

Don't worry, we'll maybe be able to use it in *checks watch* 4 years. ;-)

I think being able to pass `url` as a kwarg isn't too big a deal (and can make dynamic calling easier), but that may just be my "Python was my first programming language" knee jerking. :grin: 

Regardless, it sounds like we're agreed that the best we can do (until the middle of next decade) is modify those three callsites to pass in the previous `read_file_or_url` defaults and drop `*args` from the `read_file_or_url` function signature. @blackboxsw, are you happy with that?

Reply:
Thank you for the good and thoughtful review gentlemen.  I have since dropped *args from read_file_or_url and promoted timeout=5 and retries=10 'former-default' values up to call sites to avoid behavior changes. Since sec_between=1 default is retained in readurl, we didn't need to promote that value up to call sites. Thanks again.